File: implementations/java/provekit-lift-java-source/src/main/java/com/provekit/lift/java_source/KitDeclaration.java
// SPDX-License-Identifier: Apache-2.0

package com.provekit.lift.java_source;

import com.provekit.ir.Jcs;

final class KitDeclaration {
    static final String RPC_METHOD = "provekit.plugin.kit_declaration";

    private KitDeclaration() {}

    static Jcs.Obj toJson() {
        return Jcs.object(
            "kit", Jcs.object(
                "id", Jcs.string("java-source"),
                "language", Jcs.string("java"),
                "version", Jcs.string("0.1.0")
            ),
            "rpc", Jcs.object(
                "methods", Jcs.array(
                    rpcMethod("initialize", true),
                    rpcMethod(RPC_METHOD, true),
                    rpcMethod("lift", true),
                    rpcMethod("shutdown", false)
                )
            ),
            "proofResolution", Jcs.object("strategy", Jcs.string("maven")),
            "effectKinds", Jcs.array(Jcs.string("concept:panic-freedom")),
            "effectLeaves", Jcs.array(
                Jcs.object(
                    "surface", Jcs.string("java-source"),
                    "local", Jcs.string("java:throw"),
                    "concept", Jcs.string("concept:panic-freedom.leaf.runtime-failure-site")
                )
            ),
            "guardPredicates", Jcs.array(),
            "controlCarriers", Jcs.array(),
            "residueCategories", Jcs.array()
        );
    }

    private static Jcs.Obj rpcMethod(String name, boolean required) {
        return Jcs.object(
            "name", Jcs.string(name),
            "required", Jcs.bool(required)
        );
    }
}

File: implementations/go/cmd/provekit-lift-go-verify/kit_declaration.go
package main

const kitDeclarationRPCMethodName = "provekit.plugin.kit_declaration"

func kitDeclarationResult() map[string]any {
	return map[string]any{
		"kit": map[string]any{
			"id":       "go",
			"language": "go",
			"version":  "0.1.0",
		},
		"rpc": map[string]any{
			"methods": []any{
				map[string]any{"name": "initialize", "required": true},
				map[string]any{"name": kitDeclarationRPCMethodName, "required": true},
				map[string]any{"name": "lift", "required": true},
				map[string]any{"name": "shutdown", "required": false},
			},
		},
		"proofResolution": map[string]any{"strategy": "go-mod"},
		"effectKinds":     []any{"concept:panic-freedom"},
		"effectLeaves": []any{
			map[string]any{
				"surface": "go",
				"local":   "go:panic",
				"concept": "concept:panic-freedom.leaf.runtime-failure-site",
			},
		},
		"guardPredicates":   []any{},
		"controlCarriers":   []any{},
		"residueCategories": []any{},
	}
}
